Issue/add username changer

theck13 opened the pull request:
Add a method for updating the username of a WordPress.com account, `pushUsername`.  It requires the `username` and `action` parameters using the `/me/username` authenticated endpoint.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/account/AccountRestClient.java:
```diff
@@ -624,6 +668,26 @@ private AccountPushSocialResponsePayload volleyErrorToAccountSocialResponsePaylo
         return payload;
     }
 
+    private AccountPushUsernameResponsePayload volleyErrorToAccountUsernameResponsePayload(VolleyError error) {
```
I am not really sure why we might need this. For this particular request, I'd guess something like this: https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/post/PostRestClient.java#L78-L80 should work, but I am not 100% sure.

Once we add some tests for this (comment incoming about that :) ) we should try this out and only follow this explicit pattern if necessary.

theck13 replied:
Ahh, I didn't know about that pattern.  I was just following [these two methods](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/account/AccountRestClient.java#L587-L625).  I think [`volleyErrorToAccountResponsePayload`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/account/AccountRestClient.java#L587-L602) can be changed to the pattern you suggested too.  I updated `pushUsername` in 8d6f9756f65d81935321af8c130467dd1e78cd65.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/AccountStore.java:
```diff
@@ -161,6 +171,9 @@ public OnSocialChanged(@NonNull AccountPushSocialResponsePayload payload) {
         }
     }
 
+    public static class OnUsernameChanged extends OnChanged<AccountUsernameError> {
```
Similar to my previous comment, we should have information about which `account` and `username` this even is related to.

oguzkocer replied:
Let's skip adding the `account` to the event as it looks like we never do it. However, we should have the `username` here and probably the `AccountUsernameActionType` so we know which action is taken in the callback.

theck13 commented:
> We probably should have a property here for which `username` and `account` this response is for.

I can add the `username` to the `AccountPushUsernameResponsePayload`, I'm not sure about the `account`.

> Most important of all, we should add some tests for both successful and error paths

I added the tests in 8f2687d726ad4a0e9728cdf0f90b2d748cd386e1.  I created a [new issue](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/656) for testing the success cases as we discussed.

I also added the `AccountUsernameActionType` enumeration to simplify creating the payload in the client app in 56f9dce379680f4e881459703220ff9f8a5bc3e1 and f74c86cf40b6b4375490f857c92a8a2c5ef117b8.  Thanks for the suggestion!

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -39,7 +42,9 @@
         AUTH_EMAIL_ERROR_INVALID,
-        AUTH_EMAIL_ERROR_NO_SUCH_USER
+        AUTH_EMAIL_ERROR_NO_SUCH_USER,
+        USERNAME_ERROR_GENERIC,
```
How do you feel about renaming this to `CHANGE_USERNAME_ERROR_GENERIC` to be more specific? Also the next one to `CHANGE_USERNAME_ERROR_INVALID`?

theck13 replied:
I updated those values in b1f89b5ca20fbf603b49ec020a48d26422efb933.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -152,6 +157,28 @@ public void testWPComPostPrimarySiteIdNoChange() throws InterruptedException {
         assertEquals(newValue, String.valueOf(mAccountStore.getAccount().getPrimarySiteId()));
     }
 
+    public void testWPComUsernameError() throws InterruptedException {
```
I'd suggest separating this into 2 different tests. Unless there is a continuation between tests, I think it's always best to treat them separately. How about: `testChangeWPComUsernameGenericError` & `testChangeWPComUsernameInvalidActionError`?

theck13 replied:
I updated those methods in d78567a1da754aff09e38de67688bde0ca759975.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -152,6 +157,28 @@ public void testWPComPostPrimarySiteIdNoChange() throws InterruptedException {
         assertEquals(newValue, String.valueOf(mAccountStore.getAccount().getPrimarySiteId()));
     }
 
+    public void testWPComUsernameError() throws InterruptedException {
```
The test fails with `AUTHORIZATION_REQUIRED` error. We need something like this before starting the test to authenticate the user first:

```
mNextEvent = TestEvents.AUTHENTICATE;
authenticate(BuildConfig.TEST_WPCOM_USERNAME_TEST1, BuildConfig.TEST_WPCOM_PASSWORD_TEST1);
```

theck13 replied:
I added the authentication check in 7f32cf361e2e1cee924ffd1d865f7ad22246cf19.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -276,6 +303,32 @@ public void onAuthEmailSent(OnAuthEmailSent event) {
+                    assertEquals(mNextEvent, TestEvents.USERNAME_ERROR_INVALID);
+                    mCountDownLatch.countDown();
+                    break;
+                case INVALID_INPUT:
```
We need to throw an error here if we don't expect it to happen. Otherwise the test will pass if we ever get `INVALID_INPUT` error when we expect the call to actually succeed. (Not in it's current form, in it's current form it'll just hang because we never call `mCountDownLatch.countDown();`, but I assume we'll cover the success path)

theck13 replied:
I added the error statement in 006d6242e40e3b05038da66b4c03e695e45824fb.

theck13 commented:
I added the `username` and `AccountUsernameActionType` variables in 9355bd4fa040a7b105d0b34c6f7d6abf17eb9ac3.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -40,8 +43,11 @@
+        AUTH_EMAIL_ERROR_NO_SUCH_USER,
         AUTH_EMAIL_ERROR_USER_EXISTS,
-        AUTH_EMAIL_ERROR_NO_SUCH_USER
+        AUTH_EMAIL_ERROR_NO_SUCH_USER,
```
We now have double `AUTH_EMAIL_ERROR_NO_SUCH_USER`.

theck13 replied:
I removed the duplicated item in 5333ea6cb6b04de227ec6a2fbeafe548e6f764b4.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -154,6 +160,48 @@ public void testWPComPostPrimarySiteIdNoChange() throws InterruptedException {
         assertEquals(newValue, String.valueOf(mAccountStore.getAccount().getPrimarySiteId()));
     }
 
+    public void testWPComUsernameGenericError() throws InterruptedException {
```
I think we can improve on the test names as suggested [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/655#discussion_r158392774). Without the mention of "username change", just reading the test name doesn't really give enough context which can be really important because otherwise we'd have to read to whole test to understand what's going on.

I know I initially suggested `testChangeWPComUsernameGenericError`, but `testChangeWPComUsernameWithSameUsernameError` might be better, because generic error is again doesn't give enough context. The test naming should say what we are doing rather than what we expect which I probably am at fault as well in some of my recent tests.

These are just my thoughts on it, but I am totally OK if you want to keep it as is or change it to something else.

theck13 replied:
Since we're just testing one situation for all the `GENERIC_ERROR` cases, I think `testChangeWPComUsernameGenericError` is good.  If we add every `GENERIC_ERROR` case, then the `testChangeWPComUsernameWithSameUsernameError` convention is better.

I originally missed the `Change` part of the name in your previous comment.  I updated the tests in e2d899bc5325d1462cdff7e45054bc54f84ab77d.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -154,6 +160,48 @@ public void testWPComPostPrimarySiteIdNoChange() throws InterruptedException {
         assertEquals(newValue, String.valueOf(mAccountStore.getAccount().getPrimarySiteId()));
     }
 
+    public void testWPComUsernameGenericError() throws InterruptedException {
```
This test fails with `INVALID_INPUT` error [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/issue/add-username-changer/example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java#L385). The test expects a generic error, so I guess that's not the case?

theck13 replied:
I mixed up the `GENERIC_ERROR` and `INVALID_INPUT` cases.  I updated the tests in c1a3c71f967cc38111b23b1e49422c681fc32df4.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -154,6 +160,48 @@ public void testWPComPostPrimarySiteIdNoChange() throws InterruptedException {
+        String address = mAccountStore.getAccount().getWebAddress();
+
+        PushUsernamePayload payload = new PushUsernamePayload(username,
+                AccountUsernameActionType.valueOf("invalid action"));
```
```
AccountUsernameActionType.valueOf("invalid action")
```
This gets an `IllegalArgumentException - No enum constant org.wordpress.android.fluxc.store.AccountStore.AccountUsernameActionType.invalid action` error. If this is the only way to trigger the `INVALID_ACTION` error, we can skip this test and throw an exception [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/issue/add-username-changer/example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java#L379) saying this shouldn't happen with `AccountUsernameActionType` in place.

theck13 replied:
That makes sense.  I updated the test case in 52dd1899689f78c46f3db896ee8defc11d72d083.

oguzkocer commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_AccountTest.java:
```diff
@@ -319,6 +344,29 @@ public void onAuthEmailSent(OnAuthEmailSent event) {
+                case GENERIC_ERROR:
+                    throw new AssertionError("Error should not be tested: " + event.error.type);
+                case INVALID_ACTION:
+                    throw new AssertionError("Error should not occur with action type enum: " + event.error.type);
```
Even error type is not actually the action type here. Maybe we can change the `event.error.type` to `event.type` or change the message? What do you think?

theck13 replied:
I updated the error message in 8929981ec317132aaa6aeef640e4d44516a20bb2.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/account/AccountRestClient.java:
```diff
@@ -127,6 +129,11 @@ public boolean hasUsername() {
 
+    public static class AccountPushUsernameResponsePayload extends Payload<AccountUsernameError> {
+        public AccountUsernameActionType type;
+        public String username;
```
How do you feel about having a constructor here that takes `username` and `AccountUsernameActionType`? Since we always add these to the payload, I thought it might be a nice improvement.

theck13 replied:
I added the constructor in 79903a30c901ef1edf9b7f5fb1d000d1d5a6bcbb.

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/AccountStore.java:
```diff
@@ -648,6 +719,14 @@ private void handlePushSocialCompleted(AccountPushSocialResponsePayload payload)
+        OnUsernameChanged onUsernameChanged = new OnUsernameChanged();
+        onUsernameChanged.username = payload.username;
+        onUsernameChanged.type = payload.type;
+        onUsernameChanged.error = payload.error;
```
We probably want to update the account in DB after a successful username change.

theck13 replied:
I added updating the database with the username in 90332a4a2eca851b7aed9f425d5344f486336cdd and d0e5d2fa77f2638126e62ee83b18cbfd78c3cfa1.

theck13 commented:
No problem.  Thanks for bearing with _me_!

oguzkocer commented:
Travis shows that the last commit has failed on this PR, which is not entirely correct. Before the PR was merged, Travis passed both checks.

Travis was a little slow today, so I tried to rush a build which put it at the end of queue and I think once I deleted the branch, it failed on that manully triggered build.

Just wanted to put this here for future reference.
